libtiledbsoma/src/soma/managed_query.h

```diff
@@ -805,6 +811,139 @@ class ManagedQuery {
+    _find_covered_enum_values(
+        const std::vector<std::string_view>& enum_values_as_sv,
+        ArrowSchema* index_schema,
+        ArrowArray* index_array) {
+        if (index_schema == nullptr) {
+            return std::nullopt;
+        }
+        auto user_index_type = ArrowAdapter::to_tiledb_format(
+            index_schema->format);
+        switch (user_index_type) {
+            case TILEDB_INT8:
+                return _find_covered_enum_values_aux<int8_t>(
+                    enum_values_as_sv, index_array);
+            case TILEDB_UINT8:
+                return _find_covered_enum_values_aux<uint8_t>(
+                    enum_values_as_sv, index_array);
+            case TILEDB_INT16:
+                return _find_covered_enum_values_aux<int16_t>(
+                    enum_values_as_sv, index_array);
+            case TILEDB_UINT16:
+                return _find_covered_enum_values_aux<uint16_t>(
+                    enum_values_as_sv, index_array);
+            case TILEDB_INT32:
+                return _find_covered_enum_values_aux<int32_t>(
+                    enum_values_as_sv, index_array);
+            case TILEDB_UINT32:
+                return _find_covered_enum_values_aux<uint32_t>(
+                    enum_values_as_sv, index_array);
+            case TILEDB_INT64:
+                return _find_covered_enum_values_aux<int64_t>(
+                    enum_values_as_sv, index_array);
+            case TILEDB_UINT64:
+                return _find_covered_enum_values_aux<uint64_t>(
+                    enum_values_as_sv, index_array);
+            default:
+                throw TileDBSOMAError(
+                    "Saw invalid enumeration index type when trying to extend"
+                    "enumeration");
+        }
+    }
+
+    template <typename IndexType>
+    std::optional<std::unordered_set<std::string_view>>
+    _find_covered_enum_values_aux(
+        const std::vector<std::string_view>& enum_values_as_sv,
+        ArrowArray* index_array) {
+        std::unordered_set<std::string_view> retval;
+
+        if (index_array == nullptr) {
+            // All enum values are considered covered: this is the context
+            // of our having been given a plain Arrow array of enum values
+            // and being asked to extend the array's enumeration to include
+            // them. We should honor all the provided values.
+            return std::nullopt;
+        }
+
+        std::optional<std::vector<uint8_t>>
+            opt_validities = _cast_validity_buffer(index_array);
+        if (!opt_validities.has_value()) {
+            // All enum values are considered covered: this is the context
+            // where the user has given us an Arrow dictionary, and
+            // the index part is all non-null values and the Arrow library
+            // has no validity buffer at all for it.
+            return std::nullopt;
+        }
+        std::vector<uint8_t> validities = opt_validities.value();
```

Comment:
```suggestion
        std::vector<uint8_t>& validities = opt_validities.value();
```
`std::optional::value` returns one of the following:
* `T&`
* `const T&`
* `T&&`
* `const T&&`

I'm pretty sure we want `T&` - what's written here might ultimately do the right thing, but I'm not 100% sure either way.

Reply:
Definitely I want the reference -- thank you @jp!!

Reply:
https://github.com/single-cell-data/TileDB-SOMA/pull/3992/commits/77c5eca679577115bd35a401af384f8aba7d3f3e